### ISSUE TITLE
Code assist not working in XML file

### ISSUE THREAD

**Issue by mauromol** [spring-projects/sts4#415]
Trying to investigate as per https://github.com/spring-projects/sts4/issues/340#issuecomment-573036075

In my XML file where I can't get code assist to work, I try to write a bean definition and I invoke code assist in the `class` attribute value:

    <bean class="Con|" />

I invoke code assist at "|" and I get the following in the Language Server console:

```
15:20:31.291 [elastic-2] INFO  o.s.i.v.c.l.c.VscodeCompletionEngineAdapter - Starting completion handling
15:20:31.291 [elastic-2] INFO  o.s.i.v.c.l.c.CompositeLanguageServerComponents - languageId = LanguageId [xml]
15:20:31.295 [elastic-2] INFO  o.s.i.v.c.l.c.VscodeCompletionEngineAdapter - Got 0 completions
15:20:31.297 [elastic-2] INFO  o.s.i.v.c.l.c.VscodeCompletionEngineAdapter - Completion handling terminated!
```

So, it seems like some kind of triggering is performed, but 0 completions are returned. In fact, I get no suggestions at all in the editor.

In the Error Log, at the same time, I get some errors from the XMLLanguageServer contributed by the Wild Web Developer plugin from Eclipse. Can this interfere with STS4 Language Server?
For reference, the error I get is the same I get on Eclipse start which I described at https://github.com/eclipse/wildwebdeveloper/issues/366.

There are no errors from STS4 in the Error Log though.

**Comment by BoykoAlex** [comment]
Can you paste your entire XML here please? There is an issue with XML I recall that if there is some kind of header present it may not work... Think @kdvolder discovered it once but I just can't remember exactly what that was unfortunately...

**Comment by BoykoAlex** [comment]
@mauromol one silly thing... did you enable **Spring XML File config support** in the preferences (under Language Servers -> Spring Language Servers -> Spring Boot Language Server -> Spring XML Config)

**Comment by mauromol** [comment]
> @mauromol one silly thing... did you enable **Spring XML File config support** in the preferences (under Language Servers -> Spring Language Servers -> Spring Boot Language Server -> Spring XML Config)

Yes I did.

The full XML contains some sensitive data, I should clean it before submitting. If you meanwhile have any hint on what part of it could be causing this, I may try to do some tests.

**Comment by BoykoAlex** [comment]
Very likely this is just a side effect of not having proper classpath - issue #413.

**Comment by mauromol** [comment]
Hi Alex,
just a note to notify that, as you can read at https://github.com/spring-projects/sts4/issues/413#issuecomment-596625804, it seems like the content assist is not working even when I do not get the timeout exception described in #413. I mean, in that log you can see that for my `shop-backend` project it detects 150 entries, but still it returns an empty suggestion list.
I'm trying to fire content assist in `class` attribute value of a `<bean>` declaration.

**Comment by BoykoAlex** [comment]
It'd be great to look at this XML file... What kind of CA is that by the way? Inside `class` `name` or `ref` attribute? Does it matter which attribute? Or is it no CA for any of them?

Here is the example XML where CA and various Spring Beans XML feature work:
```
<?xml version="1.0" encoding="UTF-8"?>
<beans xmlns="http://www.springframework.org/schema/beans"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://www.springframework.org/schema/beans https://www.springframework.org/schema/beans/spring-beans.xsd">
	
	<!-- Root Context: defines shared resources visible to all other web components -->
	<bean id="simpleObj" class="u.t.r.SimpleObj"></bean>

	<bean id="testBean" class="u.t.r.TestBean"
		scope="prototype">
		<property name="age" value="10" />
		<property name="simple" ref="simpleObj"></property>
	</bean>
		
</beans>
```

Is your DOM structure any different from the one above? Any extra or missing attributes or elements?

**Comment by mauromol** [comment]
From https://github.com/spring-projects/sts4/issues/413#issuecomment-596631398:

> It'd be great to look at this XML file...

If you think it's possible, I can mail you this XML file privately. Anyway, from a first look, it seems like its structure is like the one you mention. I just have some custom namespace tags interleaving standard `<bean>` declarations.

> What kind of CA is that by the way? Inside class name or ref attribute?

The referenced log was produced by invoking code assist inside `class` attribute of `<bean>`.

> Does it matter which attribute? Or is it no CA for any of them?

I can't get CA to work at any place of a Spring XML file. If you reconstruct the whole story from #340 you'll see that I was never able to make STS4 XML content assist to work on this project, so that I worried whether STS4 was working at all in my environment.

**Comment by BoykoAlex** [comment]
It'd be even better to have the whole problematic project but we can start with just the XML. Yes, please, feel free to e-mail me directly. (aboyko at pivotal dot io)

**Comment by BoykoAlex** [comment]
I took a few XML files an placed them into my sample project and i can see that the CA is working...
Inside `class` attributes you may want to start typing the package name like `class=o^` and then invoke CA... seemed like CA didn't work on an empty string value...
The CA inside `<property name="^" value="true" />` should work well though... I get all kinds of property names for a bean with `class="org.springframework.beans.factory.config.PropertyPlaceholderConfigurer"`. Are you getting anything there?
Feels like i may need to experiment with this in a Gradle project...

**Comment by BoykoAlex** [comment]
Tried the same XML files in the Gradle boot project with war packaging. They seemed to work fine.
I have added some logging to the Spring Boot LS process that may help figuring out what is going to wrong.
Please grab the latest STS4 for Eclipse. Either the distro CI-B2184 (or later) or update from the nightly: https://dist.springsource.com/snapshot/TOOLS/sts4/nightly/e4.15
Enable Spring Boot Debug Console via Preferences -> Language Servers -> Spring Language Servers preference page
Enable Spring XML Config files support via Preferences -> Language Servers -> Spring Language Servers -> Spring Boot Language Server -> Spring XML Config preference page

Import your Gradle multiproject in the workspace open XML file and invoke CA in the `class` attribute (just ensure you have non-empty prefix, i.e. `org`) and then in the property `name` attribute (the class for the bean should be specified correctly in this case)

Record portions of log for each CA invocation string from: `Stating calculating completions for...` and ending with `Completion item resolve request terminated.`. Please paste or attach these log chunks here and then we'll try to analyze them and determine what to do next.

**Comment by mauromol** [comment]
Hi Alex,
I'm currently using B2183, however if you have a look at #413 it seems like Martin has made it work in some way, you'll find some logs there, even if they do not contain the additional logging you just put in.
Let me know if you still need testing from me. I don't know exactly what Martin did, perhaps you can sync with him to clarify what was wrong with regards to CA...

**Comment by mauromol** [comment]
Hi @BoykoAlex, if you and @martinlippert agree that this issue was indeed caused by what has been done and fixed along #413, I think we can close this. If I'll have other problems with code assist in XML files I will report back. My original issue (code assist not working at all with my XML files) seems to be fixed in the latest STS4 integration builds.

**Comment by BoykoAlex** [comment]
Oh, yes, lets indeed close this one.
